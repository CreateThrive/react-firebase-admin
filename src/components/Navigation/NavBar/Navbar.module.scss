.avatar{
    font-size: 28pt;
    font-weight: bold;
    margin-top: -0.1rem;
}

