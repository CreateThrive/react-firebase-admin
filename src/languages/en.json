{
  "Home.home": "Home",
  "Home.content": "Home content",
  "Login.login": "Login",
  "Login.setNewPassword": "Set your new password",
  "Login.email": "E-mail Address",
  "Login.password": "Password",
  "Login.setPassword": "Set Password",
  "Login.forgotPassword": "Forgot Password?",
  "Login.facebook": "Continue with Facebook",
  "Login.google": "Continue with Google",
  "Login.microsoft": "Continue with Microsoft",
  "Login.safePassword": "Safe password",
  "Login.unsafePassword": "Unsafe password",
  "Login.invalidEmail": "Invalid email",
  "Login.invalidPassword": "Invalid password",
  "NotFound.404": "Error 404: page not found",
  "NotFound.url": "The requested URL {url} was not found",
  "NotFound.back": "Go Back",
  "Profile.profile": "Profile",
  "ChangePassword.samePassword": "The new password and the current one cannot be the same",
  "ChangePassword.changePassword": "Change Password",
  "ChangePassword.currentPassword": "Current Password",
  "ChangePassword.newPassword": "New Password",
  "ChangePassword.confirmPassword": "Confirm Password",
  "ChangePassword.submits": "Submit",
  "ChangePassword.safePassword": "Safe password",
  "ChangePassword.insecurePassword": "Insecure password",
  "ChangePassword.matchPassword": "Passwords match",
  "ChangePassword.notMatchPassword": "Passwords do not match",
  "ChangePassword.invalidPassword": "Password must be at least six characters long",
  "Section.section": "Section",
  "Section.content": "Section content",
  "ResetPassword.recovery": "Password Recovery",
  "ResetPassword.recoverEmail": "We have sent you an email to {mail} so you can recover your account.",
  "ResetPassword.email": "E-mail Address",
  "ResetPassword.emailRegistration": "E-mail used for registration",
  "ResetPassword.resetLink": "Send Reset Link",
  "ResetPassword.back": "Back",
  "Submenu.submenu": "Submenu",
  "Submenu.content": "Submenu content",
  "User.editUser": "Edit User",
  "User.newUser": "New User",
  "Users.name": "Name",
  "Users.email": "Email",
  "Users.location": "Location",
  "Users.admin": "Admin",
  "Users.created": "Created",
  "Users.delete": "Delete",
  "Users.confirm": "Confirm action",
  "Users.permDelete": "This will permanently delete the user. Action can not be undone.",
  "Users.cancel": "Cancel",
  "Users.users": "Users",
  "Users.newUser": "New User",
  "Users.search": "Search:",
  "Aside.home": "Home",
  "Aside.users": "Users",
  "Aside.teams": "Teams",
  "Aside.dropdownMenu": "Dropdown Menu",
  "Aside.submenu1": "Submenu 1",
  "Aside.submenu2": "Submenu 2",
  "NavBar.profile": "Profile",
  "NavBar.logOut": "Log Out",
  "NavBar.Language": "English",
  "Table.perPage": " per page",
  "UserForm.invalidEmail": "Invalid E-mail",
  "UserForm.userInfo": "User Information",
  "UserForm.email": "E-mail",
  "UserForm.name": "Name",
  "UserForm.location": "Location",
  "UserForm.admin": "Admin",
  "UserForm.created": "Created",
  "UserForm.logo": "Logo",
  "UserForm.pickAnotherFile": "Pick another file",
  "UserForm.pickFile": "Pick a file",
  "UserForm.submit": "Submit",
  "UserForm.goBack": "Go Back",
  "UserForm.userPreview": "User Preview",
  "Teams.teams": "Teams",
  "TeamsCard.allTeams": "Available teams",
  "TeamsCard.addTeam": "Add Team",
  "TeamsCard.allUsers": "Team users",
  "TeamsCard.addUser": "Add User",
  "TeamsCard.delete": "Delete",
  "TeamsCard.searchBy": "Search by name",
  "TeamsCard.confirm": "Confirm action",
  "TeamsCard.permDeleteTeam": "This will permanently delete the team. Action can not be undone.",
  "TeamsCard.permDeleteUser": "This will permanently delete the user from the team. Action can not be undone.",
  "TeamsCard.cancel": "Cancel",
  "TeamModal.teamInformation": "Team Information",
  "TeamModal.addTeam": "Add Team",
  "TeamModal.updateTeam": "Update Team",
  "TeamModal.cancel": "Cancel",
  "TeamModal.name": "Name",
  "TeamModal.description": "Description",
  "users.UserNotAvailable": "User not available",
  "auth/email-already-exists": "Email already in use",
  "auth/invalid-email": "Email is invalid",
  "auth/wrong-password": "Invalid credentials",
  "auth/user-disabled": "User disabled",
  "auth/too-many-requests": "Too many attempts made, try again later",
  "auth/expired-action-code": "The invitation link has expired, get in touch with your administrator",
  "auth/invalid-action-code": "The invitation link has expired, get in touch with your administrator",
  "storage/quota-exceeded": "Internal server error, get in touch with your administrator",
  "storage/unauthenticated": "Unauthenticated, please authenticate and try again",
  "storage/unauthorized": "Unauthorized, you are not authorized to perform this action",
  "auth/account-exists-with-different-credential": "Email on another sign in method",
  "auth/popup-blocked": "Pop up blocked by the browser, please enable pop ups for this page",
  "auth/popup-closed-by-user": "Operation cancelled, pop up was closed",
  "auth/cancelled-popup-request": "Too many pop ups, just one pop up is allowed",
  "utils.default": "Unknown error, get in touch with your administrator",
  "ErrorMessage.defaultMessage": "This field is required",
  "invalidEmail": "Invalid E-mail"
}
