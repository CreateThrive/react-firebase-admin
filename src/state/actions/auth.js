import { createAction } from 'redux-act';
import { toastr } from 'react-redux-toastr';

import { firebaseError, FIREBASE_RESPONSE } from 'utils';
import firebase from 'firebase.js';
import { clearUsersDataLogout } from './users';

export const AUTH_SIGN_IN_INIT = createAction('AUTH_SIGN_IN_INIT');
export const AUTH_SIGN_IN_FAIL = createAction('AUTH_SIGN_IN_FAIL');

export const AUTH_RESTORE_SESSION_INIT = createAction(
  'AUTH_RESTORE_SESSION_INIT'
);
export const AUTH_RESTORE_SESSION_SUCCESS = createAction(
  'AUTH_RESTORE_SESSION_SUCCESS'
);
export const AUTH_RESTORE_SESSION_FAIL = createAction(
  'AUTH_RESTORE_SESSION_FAIL'
);

export const AUTH_LOGOUT_INIT = createAction('AUTH_LOGOUT_INIT');
export const AUTH_LOGOUT_SUCCESS = createAction('AUTH_LOGOUT_SUCCESS');

export const AUTH_SET_PASSWORD_INIT = createAction('AUTH_SET_PASSWORD_INIT');
export const AUTH_SET_PASSWORD_SUCCESS = createAction(
  'AUTH_SET_PASSWORD_SUCCESS'
);
export const AUTH_SET_PASSWORD_FAIL = createAction('AUTH_SET_PASSWORD_FAIL');

export const AUTH_RESET_PASSWORD_INIT = createAction(
  'AUTH_RESET_PASSWORD_INIT'
);
export const AUTH_RESET_PASSWORD_SUCCESS = createAction(
  'AUTH_RESET_PASSWORD_SUCCESS'
);
export const AUTH_RESET_PASSWORD_FAIL = createAction(
  'AUTH_RESET_PASSWORD_FAIL'
);

export const AUTH_CLEAN_UP = createAction('AUTH_CLEAN_UP');

export const AUTH_FETCH_USER_DATA_INIT = createAction(
  'AUTH_FETCH_USER_DATA_INIT'
);
export const AUTH_FETCH_USER_DATA_SUCCESS = createAction(
  'AUTH_FETCH_USER_DATA_SUCCESS'
);
export const AUTH_FETCH_USER_DATA_FAIL = createAction(
  'AUTH_FETCH_USER_DATA_FAIL'
);

export const AUTH_CHANGE_PASSWORD_INIT = createAction(
  'AUTH_CHANGE_PASSWORD_INIT'
);
export const AUTH_CHANGE_PASSWORD_SUCCESS = createAction(
  'AUTH_CHANGE_PASSWORD_SUCCESS'
);
export const AUTH_CHANGE_PASSWORD_FAIL = createAction(
  'AUTH_CHANGE_PASSWORD_FAIL'
);

export const AUTH_UPDATE_USER_DATA = createAction('AUTH_UPDATE_USER_DATA');

export const AUTH_PROVIDER_INIT = createAction('AUTH_PROVIDER_INIT');

export const AUTH_PROVIDER_SUCCESS = createAction('AUTH_PROVIDER_SUCCESS');

export const AUTH_PROVIDER_FAIL = createAction('AUTH_PROVIDER_FAIL');

export const logout = () => {
  return async dispatch => {
    dispatch(AUTH_LOGOUT_INIT());

    dispatch(clearUsersDataLogout());
    await firebase.auth().signOut();

    dispatch(AUTH_LOGOUT_SUCCESS());
  };
};

export const verifyAuth = () => {
  return dispatch => {
    firebase.auth().onAuthStateChanged(user => {
      dispatch(AUTH_RESTORE_SESSION_INIT());

      if (user !== null) {
        return dispatch(AUTH_RESTORE_SESSION_SUCCESS());
      }

      dispatch(AUTH_RESTORE_SESSION_FAIL());
      return dispatch(logout());
    });
  };
};

export const fetchUserData = () => {
  return async dispatch => {
    dispatch(AUTH_FETCH_USER_DATA_INIT());

    const { uid } = firebase.auth().currentUser;

    let user;

    try {
      user = (
        await firebase
          .database()
          .ref(`users/${uid}`)
          .once('value')
      ).val();
    } catch (error) {
      dispatch(logout());
      return dispatch(AUTH_FETCH_USER_DATA_FAIL({ error }));
    }

    if (!user) {
      return dispatch(logout());
    }

    return dispatch(
      AUTH_FETCH_USER_DATA_SUCCESS({
        id: uid,
        ...user
      })
    );
  };
};

export const checkUserData = () => {
  return (dispatch, getState) => {
    const { id } = getState().auth.userData;

    if (!id) {
      dispatch(fetchUserData());
    }
  };
};

export const auth = (email, password) => {
  return async (dispatch, getState) => {
    dispatch(AUTH_SIGN_IN_INIT());
    const { locale } = getState().preferences;
    try {
      await firebase.auth().signInWithEmailAndPassword(email, password);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_SIGN_IN_FAIL({ error: errorMessage }));
    }

    const { emailVerified } = firebase.auth().currentUser;

    if (!emailVerified) {
      const errorMessage = firebaseError(
        FIREBASE_RESPONSE.USER_DISABLED,
        locale
      );
      return dispatch(AUTH_SIGN_IN_FAIL({ error: errorMessage }));
    }

    return dispatch(fetchUserData());
  };
};

export const setPassword = (email, password, url) => {
  return async (dispatch, getState) => {
    dispatch(AUTH_SET_PASSWORD_INIT());
    const { locale } = getState().preferences;

    try {
      await firebase.auth().signInWithEmailLink(email, url);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_SET_PASSWORD_FAIL({ error: errorMessage }));
    }

    const user = firebase.auth().currentUser;

    try {
      await user.updatePassword(password);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_SET_PASSWORD_FAIL({ error: errorMessage }));
    }

    dispatch(AUTH_SET_PASSWORD_SUCCESS());

    return dispatch(fetchUserData());
  };
};

export const resetPassword = email => {
  return async (dispatch, getState) => {
    dispatch(AUTH_RESET_PASSWORD_INIT());
    const { locale } = getState().preferences;

    try {
      await firebase.auth().sendPasswordResetEmail(email);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_RESET_PASSWORD_FAIL({ error: errorMessage }));
    }

    return dispatch(AUTH_RESET_PASSWORD_SUCCESS());
  };
};

export const authCleanUp = () => dispatch => dispatch(AUTH_CLEAN_UP());

export const changeUserPassword = (currentPassword, newPassword) => {
  return async (dispatch, getState) => {
    dispatch(AUTH_CHANGE_PASSWORD_INIT());
    const { locale } = getState().preferences;

    const user = firebase.auth().currentUser;

    const { email } = user;

    const credential = firebase.auth.EmailAuthProvider.credential(
      email,
      currentPassword
    );

    try {
      await user.reauthenticateWithCredential(credential);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(AUTH_CHANGE_PASSWORD_FAIL({ error: errorMessage }));
    }

    try {
      await user.updatePassword(newPassword);
    } catch (error) {
      const errorMessage = firebaseError(error, locale);
      toastr.error('', errorMessage);
      return dispatch(AUTH_CHANGE_PASSWORD_FAIL({ error: errorMessage }));
    }

    toastr.success('', 'Password changed successfully');
    return dispatch(AUTH_CHANGE_PASSWORD_SUCCESS());
  };
};

const authWithProvider = provider => {
  return async (dispatch, getState) => {
    dispatch(AUTH_PROVIDER_INIT());
    const { locale } = getState().preferences;
    let logInData;

    try {
      logInData = await firebase.auth().signInWithPopup(provider);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_PROVIDER_FAIL({ error: errorMessage }));
    }
    const { user, additionalUserInfo } = logInData;

    const { uid } = firebase.auth().currentUser;

    const createdAt = new Date().toString();

    const { location } = additionalUserInfo.profile;

    const userData = {
      isAdmin: false,
      email: user.email,
      name: user.displayName,
      createdAt,
      logoUrl: user.photoURL,
      location: location?.name || null
    };

    let userFromDb;
    try {
      userFromDb = (
        await firebase
          .database()
          .ref(`users/${uid}`)
          .once('value')
      ).val();
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_PROVIDER_FAIL({ error: errorMessage }));
    }

    if (!userFromDb) {
      try {
        await firebase
          .database()
          .ref(`users/${uid}`)
          .set(userData);
      } catch (error) {
        const errorMessage = firebaseError(error.code, locale);
        return dispatch(AUTH_PROVIDER_FAIL({ error: errorMessage }));
      }
    }

    return dispatch(
      AUTH_PROVIDER_SUCCESS({ id: uid, ...userData, ...userFromDb })
    );
  };
};

export const authFacebook = () => {
  const provider = new firebase.auth.FacebookAuthProvider();
  provider.addScope('email');
  return authWithProvider(provider);
};

export const authGoogle = () => {
  const provider = new firebase.auth.GoogleAuthProvider();
  provider.addScope('https://www.googleapis.com/auth/user.addresses.read');
  provider.addScope('https://www.googleapis.com/auth/userinfo.email');
  return authWithProvider(provider);
};

export const authMicrosoft = () => {
  const provider = new firebase.auth.OAuthProvider('microsoft.com');
  return authWithProvider(provider);
};
