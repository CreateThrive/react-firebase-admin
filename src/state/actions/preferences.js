import { createAction } from 'redux-act';

export const PREFERENCES_SET_LOCALE = createAction('PREFERENCES_SET_LOCALE');

export const setUserLocale = locale => dispatch => {
  return dispatch(PREFERENCES_SET_LOCALE({ locale }));
};
