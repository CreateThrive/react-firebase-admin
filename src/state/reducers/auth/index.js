import { createReducer } from 'redux-act';

import {
  AUTH_SIGN_IN_INIT,
  AUTH_SIGN_IN_FAIL,
  AUTH_FETCH_USER_DATA_INIT,
  AUTH_FETCH_USER_DATA_SUCCESS,
  AUTH_FETCH_USER_DATA_FAIL,
  AUTH_LOGOUT_INIT,
  AUTH_LOGOUT_SUCCESS,
  AUTH_RESTORE_SESSION_INIT,
  AUTH_RESTORE_SESSION_SUCCESS,
  AUTH_RESTORE_SESSION_FAIL,
  AUTH_SET_PASSWORD_INIT,
  AUTH_SET_PASSWORD_SUCCESS,
  AUTH_SET_PASSWORD_FAIL,
  AUTH_RESET_PASSWORD_INIT,
  AUTH_RESET_PASSWORD_SUCCESS,
  AUTH_RESET_PASSWORD_FAIL,
  AUTH_CLEAN_UP,
  AUTH_CHANGE_PASSWORD_INIT,
  AUTH_CHANGE_PASSWORD_SUCCESS,
  AUTH_CHANGE_PASSWORD_FAIL,
  AUTH_UPDATE_USER_DATA,
  AUTH_PROVIDER_FAIL,
  AUTH_PROVIDER_INIT,
  AUTH_PROVIDER_SUCCESS
} from 'state/actions/auth';

const initialState = {
  userData: {
    id: null,
    isAdmin: null
  },
  loading: false,
  error: null,
  restoring: false,
  restoringError: null,
  restoredPassword: false,
  changedPassword: false
};

export const authReducer = createReducer(
  {
    [AUTH_SIGN_IN_INIT]: () => ({
      ...initialState,
      loading: true
    }),
    [AUTH_SIGN_IN_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [AUTH_FETCH_USER_DATA_INIT]: () => ({
      ...initialState,
      loading: true
    }),
    [AUTH_FETCH_USER_DATA_SUCCESS]: (state, payload) => ({
      ...state,
      userData: {
        id: payload.id,
        isAdmin: payload.isAdmin,
        email: payload.email,
        name: payload.name,
        location: payload.location,
        logoUrl: payload.logoUrl,
        createdAt: payload.createdAt
      },
      loading: false,
      error: null
    }),
    [AUTH_FETCH_USER_DATA_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [AUTH_LOGOUT_INIT]: () => ({ ...initialState }),
    [AUTH_LOGOUT_SUCCESS]: state => ({ ...state }),
    [AUTH_RESTORE_SESSION_INIT]: state => ({ ...state, restoring: true }),
    [AUTH_RESTORE_SESSION_SUCCESS]: state => ({
      ...state,
      restoring: false,
      restoringError: null
    }),
    [AUTH_RESTORE_SESSION_FAIL]: state => ({
      ...state,
      restoring: false,
      restoringError: true
    }),
    [AUTH_SET_PASSWORD_INIT]: state => ({ ...state, loading: true }),
    [AUTH_SET_PASSWORD_SUCCESS]: state => ({
      ...state,
      loading: false,
      error: null
    }),
    [AUTH_SET_PASSWORD_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [AUTH_RESET_PASSWORD_INIT]: () => ({
      ...initialState,
      loading: true
    }),
    [AUTH_RESET_PASSWORD_SUCCESS]: state => ({
      ...state,
      loading: false,
      error: null,
      restoredPassword: true
    }),
    [AUTH_RESET_PASSWORD_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [AUTH_CLEAN_UP]: state => ({
      ...state,
      error: null,
      changedPassword: false
    }),
    [AUTH_CHANGE_PASSWORD_INIT]: state => ({
      ...state,
      loading: true
    }),
    [AUTH_CHANGE_PASSWORD_SUCCESS]: state => ({
      ...state,
      loading: false,
      changedPassword: true
    }),
    [AUTH_CHANGE_PASSWORD_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [AUTH_UPDATE_USER_DATA]: (state, payload) => ({
      ...state,
      userData: {
        id: payload.id,
        email: state.userData.email,
        isAdmin: payload.isAdmin,
        name: payload.name,
        location: payload.location,
        logoUrl: payload.logoUrl || state.userData.logoUrl,
        createdAt: payload.createdAt
      }
    }),
    [AUTH_PROVIDER_INIT]: state => ({
      ...state,
      loading: true
    }),
    [AUTH_PROVIDER_SUCCESS]: (state, payload) => ({
      ...state,
      userData: {
        id: payload.id,
        isAdmin: payload.isAdmin,
        email: payload.email,
        name: payload.name,
        location: payload.location,
        logoUrl: payload.logoUrl,
        createdAt: payload.createdAt
      },
      error: null,
      loading: false
    }),
    [AUTH_PROVIDER_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    })
  },
  initialState
);
