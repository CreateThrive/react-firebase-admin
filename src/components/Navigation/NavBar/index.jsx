import React, { useState, useCallback } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import { availableLocales, flags } from 'utils';
import { setUserLocale } from 'state/actions/preferences';
import { useFormatMessage } from 'hooks';
import { logout } from 'state/actions/auth';
import paths from 'pages/Router/paths';
import defaultLogo from 'assets/user-default-log.svg';
import Link from '../Link';

const NavBar = ({ handleMobileToggle, asideMobileActive }) => {
  const [navMobileActive, setNavMobileActive] = useState(false);

  const { userName, logoUrl, locale } = useSelector(
    (state) => ({
      userName: state.auth.userData.name,
      logoUrl: state.auth.userData.logoUrl,
      locale: state.preferences.locale,
    }),
    shallowEqual
  );

  const dispatch = useDispatch();

  const onClickLogoutHandler = () => {
    dispatch(logout());
  };

  const onMobileToggleHandler = useCallback(() => {
    setNavMobileActive(!navMobileActive);
  }, [setNavMobileActive, navMobileActive]);

  const changeLocaleHandler = (local) => {
    dispatch(setUserLocale(local));
  };

  const locales = availableLocales.filter((local) => local !== locale);

  return (
    <nav id="navbar-main" className="navbar is-fixed-top">
      <div className="navbar-brand">
        <a
          className="navbar-item is-hidden-desktop jb-aside-mobile-toggle"
          onClick={handleMobileToggle}
        >
          <span className="icon">
            <i
              className={classNames(
                'mdi',
                'mdi-24px',
                { 'mdi-forwardburger': !asideMobileActive },
                { 'mdi-backburger': asideMobileActive }
              )}
            />
          </span>
        </a>
      </div>
      <div className="navbar-brand is-right">
        <a
          className="navbar-item is-hidden-desktop jb-navbar-menu-toggle"
          data-target="navbar-menu"
          onClick={onMobileToggleHandler}
        >
          <span className="icon">
            <i
              className={classNames(
                'mdi',
                { 'mdi-dots-vertical': !navMobileActive },
                { 'mdi-close': navMobileActive }
              )}
            />
          </span>
        </a>
      </div>
      <div
        className={classNames('navbar-menu', 'fadeIn', 'animated', 'faster', {
          'is-active': navMobileActive,
        })}
        id="navbar-menu"
      >
        <div className="navbar-end">
          <div className="navbar-item has-dropdown has-dropdown-with-icons has-divider has-user-avatar is-hoverable">
            <a className="navbar-link is-arrowless">
              <div className="is-user-avatar">
                <span>
                  <img id={locale} src={flags[locale]} alt={`${locale} flag`} />
                </span>
              </div>
              <span className="icon">
                <i className="mdi mdi-chevron-down" />
              </span>
            </a>
            <div className="navbar-dropdown">
              {locales.map((local) => (
                <a
                  onClick={() => changeLocaleHandler(local)}
                  className="navbar-item"
                  id={local}
                  key={local}
                >
                  <div className="is-user-avatar">
                    <span>
                      <img src={flags[local]} alt={`${local} flag`} />
                    </span>
                  </div>
                </a>
              ))}
            </div>
          </div>
          <div className="navbar-item has-dropdown has-dropdown-with-icons has-divider has-user-avatar is-hoverable">
            <a className="navbar-link is-arrowless">
              <div className="is-user-avatar">
                <img src={logoUrl || defaultLogo} alt="User profile" />
              </div>
              <div className="is-user-name">
                <span>{userName}</span>
              </div>
              <span className="icon">
                <i className="mdi mdi-chevron-down" />
              </span>
            </a>
            <div className="navbar-dropdown">
              <Link to={paths.PROFILE} onClick={onMobileToggleHandler}>
                <span className="icon">
                  <i className="mdi mdi-account" />
                </span>
                <span>{useFormatMessage('NavBar.profile')}</span>
              </Link>
              <hr className="navbar-divider" />
              <a
                className="navbar-item"
                id="logout"
                onClick={onClickLogoutHandler}
              >
                <span className="icon">
                  <i className="mdi mdi-logout" />
                </span>
                <span>{useFormatMessage('NavBar.logOut')}</span>
              </a>
            </div>
          </div>
        </div>
      </div>
    </nav>
  );
};

NavBar.propTypes = {
  handleMobileToggle: PropTypes.func.isRequired,
  asideMobileActive: PropTypes.bool,
};

export default NavBar;
