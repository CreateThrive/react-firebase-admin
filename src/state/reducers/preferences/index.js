import { createReducer } from 'redux-act';

import { PREFERENCES_SET_LOCALE } from 'state/actions/preferences';

const initialState = {
  locale: null
};

export const preferencesReducer = createReducer(
  {
    [PREFERENCES_SET_LOCALE]: (state, payload) => ({
      ...state,
      locale: payload.locale
    })
  },
  initialState
);
