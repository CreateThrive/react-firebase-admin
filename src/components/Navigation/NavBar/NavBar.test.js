import React from 'react';
import * as reactRedux from 'react-redux';
import { Link } from 'react-router-dom';

import en from 'assets/en.png';
import * as authActions from 'state/actions/auth';
import * as preferencesActions from 'state/actions/preferences';
import NavBar from '.';

const onHandleMobile = jest.fn();

describe('<NavBar /> rendering', () => {
  it('should render without crashing', () => {
    const { component } = shallowWithProviders(
      <NavBar handleMobileToggle={onHandleMobile} />
    )({
      auth: {
        userData: {}
      }
    });

    expect(component).toMatchSnapshot();
  });

  it('should render the links correctly', () => {
    const { component } = mountWithProviders(
      <NavBar handleMobileToggle={onHandleMobile} />
    )({
      auth: {
        userData: {}
      }
    });

    expect(component.find(Link)).toHaveLength(1);
  });
});

describe('<Bar /> actions', () => {
  const dispatchMock = jest.fn();

  beforeEach(() => {
    jest
      .spyOn(reactRedux, 'useDispatch')
      .mockImplementation(() => dispatchMock);
    jest.spyOn(authActions, 'logout').mockImplementation(jest.fn);
    jest.spyOn(preferencesActions, 'setUserLocale').mockImplementation(jest.fn);
  });

  it('should dispatch logout action when the user tries to logout', () => {
    const { component } = mountWithProviders(
      <NavBar handleMobileToggle={onHandleMobile} />
    )({
      auth: {
        userData: {}
      }
    });

    component.find('#logout').simulate('click');

    expect(authActions.logout).toHaveBeenCalled();
  });

  it('should dispatch setUserLocale action when the user tries to change language', () => {
    const { component } = mountWithProviders(
      <NavBar handleMobileToggle={onHandleMobile} />
    )({
      auth: {
        userData: {}
      }
    });

    component.find('#es').simulate('click');

    expect(preferencesActions.setUserLocale).toHaveBeenCalledWith('es');
  });

  it('should display US flag when locale is set to english', () => {
    const { component } = mountWithProviders(
      <NavBar handleMobileToggle={onHandleMobile} />
    )({
      auth: {
        userData: {}
      }
    });
    expect(component.find('#en').prop('src')).toBe(en);
  });
});
