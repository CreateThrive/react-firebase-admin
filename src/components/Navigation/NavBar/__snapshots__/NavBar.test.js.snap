// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavBar /> rendering should render without crashing 1`] = `
<DocumentFragment>
  <nav
    class="navbar is-fixed-top"
    id="navbar-main"
  >
    <div
      class="navbar-brand"
    >
      <a
        class="navbar-item is-hidden-desktop jb-aside-mobile-toggle"
      >
        <span
          class="icon"
        >
          <i
            class="mdi mdi-24px mdi-forwardburger"
          />
        </span>
      </a>
    </div>
    <div
      class="navbar-brand is-right"
    >
      <a
        class="navbar-item is-hidden-desktop jb-navbar-menu-toggle"
        data-target="navbar-menu"
      >
        <span
          class="icon"
        >
          <i
            class="mdi mdi-dots-vertical"
          />
        </span>
      </a>
    </div>
    <div
      class="navbar-menu fadeIn animated faster"
      id="navbar-menu"
    >
      <div
        class="navbar-end"
      >
        <div
          class="navbar-item has-dropdown has-dropdown-with-icons has-divider has-user-avatar is-hoverable"
        >
          <a
            class="navbar-link is-arrowless"
          >
            <div
              class="is-user-avatar"
            >
              <span>
                <img
                  alt="en flag"
                  id="en"
                  src="en.png"
                />
              </span>
            </div>
            <span
              class="icon"
            >
              <i
                class="mdi mdi-chevron-down"
              />
            </span>
          </a>
          <div
            class="navbar-dropdown"
          >
            <a
              class="navbar-item"
              id="es"
            >
              <div
                class="is-user-avatar"
              >
                <span>
                  <img
                    alt="es flag"
                    src="es.png"
                  />
                </span>
              </div>
            </a>
          </div>
        </div>
        <div
          class="navbar-item has-dropdown has-dropdown-with-icons has-divider has-user-avatar is-hoverable"
        >
          <a
            class="navbar-link is-arrowless"
          >
            <div
              class="is-user-avatar"
            >
              <img
                alt="User profile"
                src="user-default-log.jpg"
              />
            </div>
            <div
              class="is-user-name"
            >
              <span />
            </div>
            <span
              class="icon"
            >
              <i
                class="mdi mdi-chevron-down"
              />
            </span>
          </a>
          <div
            class="navbar-dropdown"
          >
            <a
              class="navbar-item"
              href="/profile"
            >
              <span
                class="icon"
              >
                <i
                  class="mdi mdi-account"
                />
              </span>
              <span>
                Profile
              </span>
            </a>
            <hr
              class="navbar-divider"
            />
            <a
              class="navbar-item"
              id="logout"
            >
              <span
                class="icon"
              >
                <i
                  class="mdi mdi-logout"
                />
              </span>
              <span>
                Log Out
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
